File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    vector<int> res;
    vector<vector<int>> adjList;
    bool dfs(int node, vector<int> &visit) {
        visit[node] = 1;
        for (auto nei: adjList[node]) {
            if (visit[nei] == 1) return false;
            else if (visit[nei] == 0) {
                if (!dfs(nei, visit)) return false;
            }
        }
        visit[node] = 2;
        res.push_back(node);
        return true;
    }

    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        adjList = vector<vector<int>>(numCourses);
        vector<int> visit(numCourses, 0);
        for (auto pre: prerequisites) {
            adjList[pre[1]].push_back(pre[0]);
        }
        for (int i = 0; i < numCourses; i++) {
            if (visit[i] == 0) {
                if(!dfs(i, visit)) return false;
            }
        }
        return true;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        visiting = set()
        visited = set()
        adjList = defaultdict(list)

        for crs, pre in prerequisites:
            adjList[pre].append(crs)

        def dfs(node):
            visiting.add(node)
            for nei in adjList[node]:
                if nei in visited:
                    continue
                if nei in visiting:
                    return False
                if not dfs(nei):
                    return False
            visiting.remove(node)
            visited.add(node)
            return True
        
        for crs in range(numCourses):
            if not dfs(crs):
                return False
        
        return True
        